File: library/src/main/java/com/getstream/sdk/chat/adapter/BaseMessageListItemViewHolder.java
package com.getstream.sdk.chat.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.getstream.sdk.chat.rest.response.ChannelState;
import com.getstream.sdk.chat.view.MessageListView;
import com.getstream.sdk.chat.view.MessageListViewStyle;

public abstract class BaseMessageListItemViewHolder extends RecyclerView.ViewHolder {

    public BaseMessageListItemViewHolder(int resId, ViewGroup parent) {
        super(LayoutInflater.from(parent.getContext()).inflate(resId, parent, false));
    }

    public abstract void bind(@NonNull Context context,
                              @NonNull ChannelState channelState,
                              @NonNull MessageListItem messageListItem,
                              @NonNull MessageListViewStyle style,
                              @NonNull MessageListView.BubbleHelper bubbleHelper,
                              @NonNull MessageViewHolderFactory factory,
                              int position);
}

File: library/src/main/java/com/getstream/sdk/chat/adapter/BaseMessageListItemViewHolder.kt
package com.getstream.sdk.chat.adapter

import android.content.Context
import android.view.View
import androidx.recyclerview.widget.RecyclerView

abstract class BaseMessageListItemViewHolder<T : MessageListItem>(
    itemView: View
) : RecyclerView.ViewHolder(itemView) {

    /**
     * Workaround to allow a downcast of the MessageListItem to T
     */
    fun bindListItem(
        messageListItem: MessageListItem,
        position: Int
    ) {
        @Suppress("UNCHECKED_CAST")
        bind(messageListItem as T, position)
    }

    protected abstract fun bind(
        messageListItem: T,
        position: Int
    )

    protected val context: Context
        get() = itemView.context
}
